### PR TITLE
Add OCI connection pool to GeoRaster driver

### PR DESCRIPTION
**Pull request by fechen123**
## What does this PR do?
Implement the OCI connection pool in the GeoRaster driver. 

## What are related issues/pull requests?
NA

## Tasklist

 - [ ] AI (Copilot or something similar) supported my development of this PR
 - [ ] Make sure code is correctly formatted (cf [pre-commit configuration](https://gdal.org/development/dev_practices.html#commit-hooks))
 - [ ] Add test case(s)
 - [ ] Add documentation
 - [ ] Updated Python API documentation (swig/include/python/docs/)
 - [ ] Review
 - [ ] Adjust for comments
 - [ ] All CI builds and checks have passed
 - [ ] ADD YOUR TASKS HERE

## Environment

Provide environment details, if relevant:

* OS:
* Compiler:

### CODE REVIEW

File: frmts/georaster/georaster_dataset.cpp [hunk 1]
```diff
@@ -92,20 +92,93 @@ GDALDataset *GeoRasterDataset::Open(GDALOpenInfo *poOpenInfo)
...
+    if (pszFetched)
+    {
+        nPoolSessionIncr = atoi(pszFetched);
+    }
+
+    //  -------------------------------------------------------------------
+    //  Create a corresponding GDALDataset object
+    //  -------------------------------------------------------------------
+
+    GeoRasterDataset *poGRD = (GeoRasterDataset *)OpenDataset(
```
Comment (rouault) [inline frmts/georaster/georaster_dataset.cpp]: please use C++ cast operators everywhere now

**Comment by rouault** [comment]
There's a crash when running autotest/gcore/misc.py::test_misc_5

**Comment by fechen123** [comment]
> There's a crash when running autotest/gcore/misc.py::test_misc_5

The crash is in gcore test. It doesn't seem to be related to what is changed in this pull request. 

Any suggestion to investigate this crash? Is there any way to reproduce it locally?

**Comment by rouault** [comment]
> The crash is in gcore test. It doesn't seem to be related to what is changed in this pull request.

This tests iterates over all drivers to check how they behave with edge case values for the arguments of GDALDriver::Create(...) method.

To run it:  ``pytest autotest/gcore/misc.py::test_misc_5``

**Comment by fechen123** [comment]
> > The crash is in gcore test. It doesn't seem to be related to what is changed in this pull request.
> 
> This tests iterates over all drivers to check how they behave with edge case values for the arguments of GDALDriver::Create(...) method.
> 
> To run it: `pytest autotest/gcore/misc.py::test_misc_5`

The test failed at creating the driver. But I couldn't see the parameters that is passed to _gdal.Driver_create() to identify what edge case this is. I am not sure if I missed anything when adding GeoRasterDriver class. Could you please help to take a look? Thanks!

**Comment by rouault** [comment]
> But I couldn't see the parameters that is passed to _gdal.Driver_create() to identify what edge case this is. I am not sure if I missed anything when adding GeoRasterDriver class.

you can run "gdb --args python3 -m pytest  autotest/gcore/misc.py::test_misc_5" to investigate where it crashes . Or add print() statement in the .py file.
I don't have bandwidth to investigate that myself

File: frmts/georaster/georaster_priv.h [hunk 2]
```diff
@@ -117,6 +117,28 @@ class GeoRasterDataset;
...
 class GeoRasterWrapper;
 
+//  ---------------------------------------------------------------------------
+//  GeoRasterDriver class definitions
+//  ---------------------------------------------------------------------------
+class GeoRasterDriver final : public GDALDriver
+{
+
+  private:
+    CPLMutex *hMutex = nullptr;
```
Comment (rouault) [inline frmts/georaster/georaster_priv.h]: using of CPLMutex is deprecated. It would be much simpler to have a ``std::mutex`` and use ``std::lock_guard oLock(oMutex)`` in scopes where it is needed

File: frmts/georaster/georaster_priv.h [hunk 3]
```diff
@@ -117,6 +117,28 @@ class GeoRasterDataset;
...
+    std::map<CPLString, OWSessionPool *> oMapSessionPool{};
```
Comment (rouault) [inline frmts/georaster/georaster_priv.h:130]: use unique_ptr to save manual deletion ?

```suggestion
    std::map<CPLString, std::unique_ptr<OWSessionPool>> oMapSessionPool{};
```

File: frmts/georaster/georaster_dataset.cpp [hunk 4]
```diff
@@ -790,6 +862,16 @@ GDALDataset *GeoRasterDataset::Create(const char *pszFilename, int nXSize,
...
                                       GDALDataType eType, char **papszOptions)
 {
+    //  -------------------------------------------------------------------
+    //  Verify georaster prefix
+    //  -------------------------------------------------------------------
+
+    if (STARTS_WITH_CI(pszFilename, "georaster:") == false &&
+        STARTS_WITH_CI(pszFilename, "geor:") == false)
+    {
+        return nullptr;
```
Comment (rouault) [inline frmts/georaster/georaster_dataset.cpp:873]:
```suggestion
        CPLError(CE_Failure, CPLE_AppDefined, "Invalid connection string");
        return nullptr;
```

File: frmts/georaster/georaster_priv.h (hunk 2 above)
Comment (fechen123) [reply frmts/georaster/georaster_priv.h]: ok, I will change to std::mutex

File: frmts/georaster/georaster_priv.h (hunk 3 above)
Comment (fechen123) [reply frmts/georaster/georaster_priv.h:130]: ok

Comment (fechen123) [reply frmts/georaster/georaster_priv.h:130]: I changed the mutex usage but decided to not using unique_ptr for now. 

Somehow it seems that my pull request commits messed up. Anyway to fix this?

Comment (rouault) [reply frmts/georaster/georaster_priv.h:130]: > Somehow it seems that my pull request commits messed up. Anyway to fix this?

try "git fetch origin master"  (assuming "origin" is the name of the remove that points to OSGeo/GDAL" and "git rebase origin/master"

Comment (fechen123) [reply frmts/georaster/georaster_priv.h:130]: > > Somehow it seems that my pull request commits messed up. Anyway to fix this?
> 
> try "git fetch origin master" (assuming "origin" is the name of the remove that points to OSGeo/GDAL" and "git rebase origin/master"

"git rebase origin/master" got error:
Auto-merging frmts/georaster/georaster_dataset.cpp
CONFLICT (content): Merge conflict in frmts/georaster/georaster_dataset.cpp
Auto-merging frmts/georaster/georaster_priv.h
CONFLICT (content): Merge conflict in frmts/georaster/georaster_priv.h
Auto-merging frmts/georaster/georaster_wrapper.cpp
CONFLICT (content): Merge conflict in frmts/georaster/georaster_wrapper.cpp
Auto-merging frmts/georaster/georasterdriver.cpp
CONFLICT (add/add): Merge conflict in frmts/georaster/georasterdriver.cpp
Auto-merging frmts/georaster/oci_wrapper.cpp
CONFLICT (content): Merge conflict in frmts/georaster/oci_wrapper.cpp
Auto-merging frmts/georaster/oci_wrapper.h
CONFLICT (content): Merge conflict in frmts/georaster/oci_wrapper.h
error: could not apply 95cb6f2495... Add the support of the OCI session pool in the GeoRaster driver.
hint: Resolve all conflicts manually, mark them as resolved with
hint: "git add/rm <conflicted_files>", then run "git rebase --continue".
hint: You can instead skip this commit: run "git rebase --skip".
hint: To abort and get back to the state before "git rebase", run "git rebase --abort".
hint: Disable this message with "git config advice.mergeConflict false"
Could not apply 95cb6f2495... Add the support of the OCI session pool in the GeoRaster driver.

Comment (rouault) [reply frmts/georaster/georaster_priv.h:130]: hum, for now, don't worry about that. While the PR looks a bit weird, upon squashing the final commit should be fine

Comment (fechen123) [reply frmts/georaster/georaster_priv.h:130]: What about my local branch, should I continue the rebase, or abort it?

Comment (rouault) [reply frmts/georaster/georaster_priv.h:130]: you can "git rebase --abort" it, and push extra commits if needed
